File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Metrics/PrometheusSink.cs
using Iaiso.Audit;

namespace Iaiso.Metrics;

/// <summary>
/// IAIso Prometheus metrics sink. Structurally typed — this module
/// doesn't depend on any specific Prometheus client library. The
/// <c>prometheus-net</c> and OpenMetrics-compatible libraries satisfy
/// these interfaces with thin adapters.
/// </summary>
public sealed class PrometheusSink : ISink
{
    /// <summary>Suggested histogram buckets for <c>iaiso_step_delta</c>.</summary>
    public static readonly double[] SuggestedHistogramBuckets =
        { 0.0, 0.01, 0.03, 0.05, 0.1, 0.2, 0.5, 1.0 };

    public interface ICounter { void Inc(); }
    public interface ICounterVec { ICounter Labels(params string[] values); }
    public interface IGauge { void Set(double v); }
    public interface IGaugeVec { IGauge Labels(params string[] values); }
    public interface IHistogram { void Observe(double v); }

    private readonly ICounterVec? _events;
    private readonly ICounter? _escalations;
    private readonly ICounter? _releases;
    private readonly IGaugeVec? _pressure;
    private readonly IHistogram? _stepDelta;

    public PrometheusSink(ICounterVec? events, ICounter? escalations, ICounter? releases,
                          IGaugeVec? pressure, IHistogram? stepDelta)
    {
        _events = events;
        _escalations = escalations;
        _releases = releases;
        _pressure = pressure;
        _stepDelta = stepDelta;
    }

    public void Emit(Event @event)
    {
        _events?.Labels(@event.Kind).Inc();
        switch (@event.Kind)
        {
            case "engine.escalation":
                _escalations?.Inc();
                break;
            case "engine.release":
                _releases?.Inc();
                break;
            case "engine.step":
                if (@event.Data.TryGetValue("pressure", out var p) && p is double pd)
                    _pressure?.Labels(@event.ExecutionId).Set(pd);
                if (@event.Data.TryGetValue("delta", out var d) && d is double dd)
                    _stepDelta?.Observe(dd);
                break;
        }
    }
}

File: IAIso-v5.0/core/iaiso-java/iaiso-metrics/src/main/java/io/iaiso/metrics/PrometheusSink.java
package io.iaiso.metrics;

import io.iaiso.audit.Event;
import io.iaiso.audit.Sink;

/**
 * IAIso Prometheus metrics sink.
 *
 * <p>Structurally typed — this module doesn't depend on any specific
 * Prometheus client library. The official {@code prometheus-simpleclient}
 * and {@code prometheus-metrics} libraries satisfy these interfaces with
 * thin adapters.
 *
 * <p>Suggested histogram buckets for {@code iaiso_step_delta}:
 * see {@link #SUGGESTED_HISTOGRAM_BUCKETS}.
 */
public final class PrometheusSink implements Sink {

    /** Suggested histogram buckets for {@code iaiso_step_delta}. */
    public static final double[] SUGGESTED_HISTOGRAM_BUCKETS =
        {0.0, 0.01, 0.03, 0.05, 0.1, 0.2, 0.5, 1.0};

    public interface Counter { void inc(); }
    public interface CounterVec { Counter labels(String... values); }
    public interface Gauge { void set(double v); }
    public interface GaugeVec { Gauge labels(String... values); }
    public interface Histogram { void observe(double v); }

    private final CounterVec events;
    private final Counter escalations;
    private final Counter releases;
    private final GaugeVec pressure;
    private final Histogram stepDelta;

    public PrometheusSink(CounterVec events, Counter escalations, Counter releases,
                          GaugeVec pressure, Histogram stepDelta) {
        this.events = events;
        this.escalations = escalations;
        this.releases = releases;
        this.pressure = pressure;
        this.stepDelta = stepDelta;
    }

    @Override
    public void emit(Event event) {
        if (events != null) events.labels(event.getKind()).inc();
        switch (event.getKind()) {
            case "engine.escalation":
                if (escalations != null) escalations.inc();
                break;
            case "engine.release":
                if (releases != null) releases.inc();
                break;
            case "engine.step":
                Object p = event.getData().get("pressure");
                if (pressure != null && p instanceof Number) {
                    pressure.labels(event.getExecutionId()).set(((Number) p).doubleValue());
                }
                Object d = event.getData().get("delta");
                if (stepDelta != null && d instanceof Number) {
                    stepDelta.observe(((Number) d).doubleValue());
                }
                break;
            default:
                // ignore
        }
    }
}
